### PR TITLE
Update babel/parser to `v7.12.17`

### PR DESCRIPTION
**Pull request by sosukesuzuki**
## Description

Closes #10308 
Closes #10365

## Checklist

- [x] I’ve added tests to confirm my change works.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

## Todo

Add tests for:

- [x] babel/babel#12686
- [x] babel/babel#12684
- [x] babel/babel#12520
- [x] babel/babel#12599
- [x] babel/babel#12562
- [x] babel/babel#12661
- [x] babel/babel#12748
- [x] babel/babel#12735
- [x] babel/babel#12775
- [x] babel/babel#12771
- [x] babel/babel#12776
- [x] babel/babel#12785

### CODE REVIEW

File: tests/typescript/interface/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -745,6 +727,67 @@ interface I {
...
+  readonly g();
+}
+
+=====================================output=====================================
+interface Foo {
+  private a();
+  public b();
+  protected c();
+  static d();
+  e();
```
Comment (sosukesuzuki) [inline tests/typescript/interface/__snapshots__/jsfmt.spec.js.snap:783]: https://github.com/prettier/prettier/issues/9760

File: tests/js/async/await-keyword.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+async function foo() {
+  function bar(x = await 2) {}
+}
+
+async (x = await 2) => {};
```
Comment (fisker) [inline tests/js/async/await-keyword.js:5]: Can you also test, Note that `flow` parse this differently.

**Prettier pr-10345**
[Playground link](https://deploy-preview-10345--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEBDAzgTymABAMwFccYBLaAiCACgEpdgAdKXA4sMigI1QCdqAHrgC8uVAHdUpGLmoAmWvWABfZqqjMM2PIJFjJ02QvrCAfA2UBuEABoQEAA6co6ZKD68I4gAp8ErlFQAG0lMVzsuXlQwAGs4GABlB2jSKABzZBheQjg7OABbLjgAE2KSgBlUdMJUNLgAMQhefNQYMnTkNEIYCFsQAAsYfKCAdX7pOHRksDgE-2lSADdpTE6wdHCQVPQ4XhhvKLSW5HxgnbsAK3QBACEo2PiE1Hy4ctS4E7PckCuBBNS0kE4ABFQgQeCfILnEDJXg7XidHhFIJ9By8VIwEakYowfrIAAcAAY7GiIDsRlEHJ00ZNdosPnYAI5g+AHRwBNDoAC0UDgJRKfV4cGZpCFB1qxyQpyh3x2+VImWysoBQNB4I+Uq+dhgqC4WJxeKQcm1UVIQQBAGEIPlJSBJgBWPqEHYAFV1AWl0MWOQAklAyrAEmB0U4AIL+hIwTBAyE7ZTKIA)

```sh
--parser babel
```

**Input:**

```jsx
async function foo() {
  function bar(x = await (2)) {}
}

async (x = await (2)) => {};
```

**Output:**

```jsx
async function foo() {
  function bar(x = await(2)) {}
}

async (x = await 2) => {};

```

File: tests/js/await/out-of-function.js [hunk 3]
```diff
@@ -0,0 +1,5 @@
+export class C {
+  p = await 0;
+}
+
+await 0;
```
Comment (fisker) [inline tests/js/await/out-of-function.js:5]: `flow` `typescript` `babel` all parse them differently.

**Prettier pr-10345**
[Playground link](https://deploy-preview-10345--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEcAeAHCAnGACMAGwEMBnUvAYT2AB0o88M8BePYgd2IEt8AKAAwBKANz0AvvXqce-YSJAAaEBAwxu0UslDFs2CBwAKuhFpTFCXAJ5blAI2zEwAazgwAyhifcoAc2Qw2ACucMpwALZ2cAAm0TEAMsR+QcS+cABiOOHEMOp+yCDEQTAQSiAAFjDhhADq5bxwpF5gcO6mvNwAbrxWBWDkZT6kcLiGjr7ZyABmFsPKAFakaABCji5u7sThcPE+cNOzoSCLaO4+voRwAIpBEPAHhHMgXtjD2AUwVhiNYNjcamUMH9YDVuNEYOVkAAOATKIEQYY1RwYApAxojTr7ZQAR1u8DGqjMhVIAFooHAYjEytg4LjuDSxqlJkgZo8jsNwtwAsF2edLjc7vsWYdlDBiHZQeDIUgAEyixzcQjnSgQcLM1CkACsZSCwwAKuKzKynp0QgBJKBxWDuX7-GAAQUt7k+lwew3E4iAA)

```sh
--parser typescript
```

**Input:**

```tsx
export class C {
  p = await (0);
}

await (0);
```

**Output:**

```tsx
export class C {
  p = await(0);
}

await 0;

```

**Review by fisker — CHANGES_REQUESTED** [review]
Let's add parenthesized await for all those cases in this PR, we don't need care how it formatted, just to have a snapshot record.

File: tests/typescript/keywords/__snapshots__/jsfmt.spec.js.snap [hunk 4]
```diff
@@ -1,14 +1,14 @@
...
 exports[`keywords.ts [babel-ts] format 1`] = `
-"Unexpected token, expected \\";\\" (4:12)
-  2 |
-  3 | module Y3 {
-> 4 |     public module Module {
-    |            ^
-  5 |         class A { s: string }
-  6 |     }
-  7 |"
+"Unexpected token, expected \\"{\\" (9:12)
```
Comment (fisker) [inline tests/typescript/keywords/__snapshots__/jsfmt.spec.js.snap:4]: Can we move code of `module Y3` a separate file?

**Comment by sosukesuzuki** [comment]
@fisker Have we the way to test with only `typescript` parser(not `babel-ts`)?

**Comment by fisker** [comment]
I think your solution is the best.

**Comment by fisker** [comment]
@sosukesuzuki The AST compare tests are failing.

**Review by fisker — APPROVED** [review]
Nice job!

**Comment by fisker** [comment]
@sosukesuzuki You already print `declare`  and `readonly` in `ClassMethod`, can you check if this still needed?
 https://github.com/prettier/prettier/blob/285c15ef08a7422dc31eedcd206680ac582ce1da/src/language-js/clean.js#L201

**Comment by sosukesuzuki** [comment]
@fisker Maybe we can remove it. But somehow all tests are passed  without it in even `main` branch.

**Comment by fisker** [comment]
IDK, I added in #9768 because tests failed.

**Pull request merged**